# Airbnb Data Deep Dive – Python + Pandas Challenge

### Table of Contents
<ul>
<li><a href="#gathering">Data Gathering</a></li>
<li><a href="#assessing">Data Assessment</a></li>
<li><a href="#cleaning">Data Cleaning</a></li>
<li><a href="#enriching">Data Enrichment</a></li>
<li><a href="#analyzing">Data Analysis</a></li>
<li><a href="#insights">Key Insights</a></li>
</ul>

In [1]:
# Importing library to be used

import pandas as pd

<a id='gathering'></a>
## Data Gathering

In [ ]:
# Loading the dataset

url = "https://data.insideairbnb.com/united-kingdom/england/london/2025-06-10/visualisations/listings.csv"

df = pd.read_csv(url)
print(df.head())

       id                                             name  host_id host_name  \
0  264776                      Huge Four Bedroom Apartment  1389063       Sue   
1  264777                            One Bedroom Apartment  1389063       Sue   
2  264778          Two Bedroom Newly Refurbished Apartment  1389063       Sue   
3  264779                Refurbished Two Bedroom Apartment  1389063       Sue   
4  264780  Spacious refurbished 2 bedroom apt with balcony  1389063       Sue   

   neighbourhood_group neighbourhood  latitude  longitude        room_type  \
0                  NaN      Lewisham  51.44306   -0.01948  Entire home/apt   
1                  NaN      Lewisham  51.44284   -0.01997  Entire home/apt   
2                  NaN      Lewisham  51.44359   -0.02275  Entire home/apt   
3                  NaN      Lewisham  51.44355   -0.02309  Entire home/apt   
4                  NaN      Lewisham  51.44333   -0.02307  Entire home/apt   

   price  minimum_nights  number_of_reviews 

<a id='assessing'></a>
## Data Assessment

In [115]:
#Concise summary of the dataframe

df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 96651 entries, 0 to 96650
Data columns (total 18 columns):
 #   Column                          Non-Null Count  Dtype  
---  ------                          --------------  -----  
 0   id                              96651 non-null  int64  
 1   name                            96651 non-null  object 
 2   host_id                         96651 non-null  int64  
 3   host_name                       96611 non-null  object 
 4   neighbourhood_group             0 non-null      float64
 5   neighbourhood                   96651 non-null  object 
 6   latitude                        96651 non-null  float64
 7   longitude                       96651 non-null  float64
 8   room_type                       96651 non-null  object 
 9   price                           62684 non-null  float64
 10  minimum_nights                  96651 non-null  int64  
 11  number_of_reviews               96651 non-null  int64  
 12  last_review                     

In [116]:
# Displaying descriptive statistics

df.describe()

id       host_id  neighbourhood_group      latitude  \
count  9.665100e+04  9.665100e+04                  0.0  96651.000000   
mean   6.526020e+17  2.091790e+08                  NaN     51.509818   
std    5.708808e+17  2.141266e+08                  NaN      0.048945   
min    1.391300e+04  2.594000e+03                  NaN     51.295937   
25%    2.955518e+07  2.673176e+07                  NaN     51.484240   
50%    8.123206e+17  1.128684e+08                  NaN     51.513791   
75%    1.197378e+18  4.063762e+08                  NaN     51.539099   
max    1.439673e+18  7.001298e+08                  NaN     51.682630   

          longitude         price  minimum_nights  number_of_reviews  \
count  96651.000000  62684.000000    96651.000000       96651.000000   
mean      -0.127087    213.366058        5.429504          20.891734   
std        0.100853    860.901557       23.315086          49.922266   
min       -0.496760      6.000000        1.000000           0.000000   
25%       -0.189060     75.000000        1.000000           0.000000   
50%       -0.126990    135.000000        2.000000           4.000000   
75%       -0.067880    225.000000        4.000000          19.000000   
max        0.278960  74100.000000     1125.000000        1855.000000   

       reviews_per_month  calculated_host_listings_count  availability_365  \
count       71487.000000                    96651.000000      96651.000000   
mean            0.958877                       16.389370        139.697365   
std             1.282595                       53.299577        137.426817   
min             0.010000                        1.000000          0.000000   
25%             0.150000                        1.000000          0.000000   
50%             0.500000                        2.000000         93.000000   
75%             1.230000                        8.000000        270.000000   
max            38.410000                      495.000000        365.000000   

       number_of_reviews_ltm  license  
count           96651.000000      0.0  
mean                5.634665      NaN  
std                11.951389      NaN  
min                 0.000000      NaN  
25%                 0.000000      NaN  
50%                 0.000000      NaN  
75%                 6.000000      NaN  
max               355.000000      NaN

In [117]:
# Checking for missing values

df.isnull().sum()

id                                    0
name                                  0
host_id                               0
host_name                            40
neighbourhood_group               96651
neighbourhood                         0
latitude                              0
longitude                             0
room_type                             0
price                             33967
minimum_nights                        0
number_of_reviews                     0
last_review                       25164
reviews_per_month                 25164
calculated_host_listings_count        0
availability_365                      0
number_of_reviews_ltm                 0
license                           96651
dtype: int64

In [118]:
# Checking for duplicate values

df.duplicated().value_counts()

False    96651
Name: count, dtype: int64

In [119]:
# Checking for unusual data types

df.infer_objects().dtypes

id                                  int64
name                               object
host_id                             int64
host_name                          object
neighbourhood_group               float64
neighbourhood                      object
latitude                          float64
longitude                         float64
room_type                          object
price                             float64
minimum_nights                      int64
number_of_reviews                   int64
last_review                        object
reviews_per_month                 float64
calculated_host_listings_count      int64
availability_365                    int64
number_of_reviews_ltm               int64
license                           float64
dtype: object

<a id='cleaning'></a>
## Data Cleaning

In [ ]:
# Converting price column to float

df['price'] = df['price'].replace('[\$,]', '', regex=True).astype(float)

<>:3: SyntaxWarning: invalid escape sequence '\$'
<>:3: SyntaxWarning: invalid escape sequence '\$'
C:\Users\Ufuoma\AppData\Local\Temp\ipykernel_1860\3254760954.py:3: SyntaxWarning: invalid escape sequence '\$'
  df['price'] = df['price'].replace('[\$,]', '', regex=True).astype(float)


In [123]:
# Parsing date columns into datetime objects

df['last_review'] = pd.to_datetime(df['last_review'], errors='coerce')
df['last_review'].dtype

dtype('<M8[ns]')

In [124]:
# Handling missing values
df['reviews_per_month'] = df['reviews_per_month'].fillna(0)
df['host_name'] = df['host_name'].fillna('Unknown')
df['neighbourhood_group'] = df['neighbourhood_group'].fillna('Not Specified')
df['last_review'] = df['last_review'].fillna('N/A')
df['license'] = df['license'].fillna('N/A')

In [125]:
# Removing listings with zero availability and price

df = df[(df['availability_365'] > 0) & (df['price'] > 0)]

df.isnull().sum()

id                                0
name                              0
host_id                           0
host_name                         0
neighbourhood_group               0
neighbourhood                     0
latitude                          0
longitude                         0
room_type                         0
price                             0
minimum_nights                    0
number_of_reviews                 0
last_review                       0
reviews_per_month                 0
calculated_host_listings_count    0
availability_365                  0
number_of_reviews_ltm             0
license                           0
dtype: int64

<a id='enriching'></a>
## Data Enrichment

In [126]:
# Creating a new column (price_per_booking)

df['price_per_booking'] = df['price'] * df['minimum_nights']


In [127]:
# Categorizing the availability column

def availability_category(x):
    if x > 300:
        return 'Full-time'
    elif 100 >= x <= 300:
        return 'Part-time'
    else:
        return 'Rare'

df['availability_category'] = df['availability_365'].apply(availability_category)

In [ ]:
#Re-checking for duplicates

df.isnull().sum()

id                                0
name                              0
host_id                           0
host_name                         0
neighbourhood_group               0
neighbourhood                     0
latitude                          0
longitude                         0
room_type                         0
price                             0
minimum_nights                    0
number_of_reviews                 0
last_review                       0
reviews_per_month                 0
calculated_host_listings_count    0
availability_365                  0
number_of_reviews_ltm             0
license                           0
price_per_booking                 0
availability_category             0
dtype: int64

<a id='analyzing'></a>
## Data Analysis

This section will be addressing the below questions:

* What are the top 10 most expensive neighborhoods by average price?
* What’s the average availability and price by room type?
* Which host has the most listings?
* How does average price vary across different boroughs or districts?
* How many listings have never been reviewed?

In [129]:
# What are the top 10 most expensive neighborhoods by average price?

exp_neighborhoods = df.groupby('neighbourhood')['price'].mean().sort_values(ascending=False).head(10)
exp_neighborhoods

neighbourhood
City of London            379.090909
Lambeth                   371.780972
Kensington and Chelsea    362.972805
Westminster               343.499335
Camden                    229.625084
Islington                 218.462021
Hammersmith and Fulham    193.441929
Wandsworth                189.752249
Richmond upon Thames      184.709962
Brent                     169.960661
Name: price, dtype: float64

In [130]:
# What’s the average availability and price by room type?

room_type_stats = df.groupby('room_type')[['availability_365', 'price']].mean().round(2).sort_values('price',ascending=False)
room_type_stats

availability_365   price
room_type                                
Hotel room                 239.22  310.97
Entire home/apt            202.28  256.60
Private room               216.12  120.02
Shared room                261.08   83.64

In [139]:
# Which host has the most listing?

top_host = df[['host_name', 'calculated_host_listings_count']].sort_values('calculated_host_listings_count', ascending=False).head(1)
top_host

host_name  calculated_host_listings_count
87926  LuxurybookingsFZE                             495

In [144]:
# How does average price vary across different boroughs or districts?

avg_price_per_district = df.groupby('neighbourhood')[['price']].mean().round(2).sort_values('price', ascending=False)
avg_price_per_district

price
neighbourhood                 
City of London          379.09
Lambeth                 371.78
Kensington and Chelsea  362.97
Westminster             343.50
Camden                  229.63
Islington               218.46
Hammersmith and Fulham  193.44
Wandsworth              189.75
Richmond upon Thames    184.71
Brent                   169.96
Southwark               164.21
Tower Hamlets           161.76
Hackney                 159.86
Merton                  158.42
Newham                  146.40
Greenwich               145.84
Haringey                143.33
Ealing                  138.54
Hounslow                138.12
Barnet                  135.97
Havering                127.38
Enfield                 125.78
Kingston upon Thames    124.49
Lewisham                117.36
Bromley                 117.35
Redbridge               117.27
Waltham Forest          112.81
Barking and Dagenham    108.15
Hillingdon              107.27
Bexley                  106.01
Croydon                 105.17
Harrow                   95.96
Sutton                   92.32

In [ ]:
# How many listings have never been reviewed?

never_been_reviewed_listings = (df['number_of_reviews'] == 0).sum()
print(f"The number of unreviewed listings is {never_been_reviewed_listings}")


The number of unreviewed listings is 14853


<a id='insights'></a>
## Key Insights

The analysis of London Airbnb listings shows that prices are heavily influenced by **location**, with central areas such as the **City of London** and **Lambeth** recording the **highest** average rates (above £370), while outer boroughs like Sutton and Harrow are far more affordable with average price below £100. 

Among room types, **hotel rooms** have the **highest** average price at about £310.97 per night, followed by entire homes or apartments at roughly £256.60. Private rooms and shared rooms are much more affordable, averaging £120.02 and £83.64 respectively. **Availability also varies**, with shared rooms and hotel rooms tending to be open for more days in a year compared to entire homes/apartments and private rooms.

The most active host based on this analysis is **LuxurybookingsFZE**, managing 495 listings, which indicates a strong professional or commercial presence in the London Airbnb market.

Interestingly, **14,853 listings have never been reviewed**, hinting at low guest engagement or an oversupply of properties in some areas.